### DISCUSSION TITLE
crypto_pwhash_str not working on iOS

### DISCUSSION THREAD

**Discussion by jdelromano** [jedisct1/libsodium#1432] [Q&A]
This code runs well on Windows, Windows Arm, Linux, OSX and Android.
But it fails on iOS, when using the .xcframework built with the ./dist-build/apple-xcframework.sh script.
```c++
char PASSWORD[] = "Just some text to make a test";
char hashed_password[crypto_pwhash_STRBYTES];
if (crypto_pwhash_str(
        hashed_password, PASSWORD, strlen(PASSWORD),
        crypto_pwhash_OPSLIMIT_SENSITIVE, crypto_pwhash_MEMLIMIT_SENSITIVE) != 0) {
    std::cout << "out of memory";
    return false;
}
if (crypto_pwhash_str_verify(
        hashed_password, PASSWORD, strlen(PASSWORD)) != 0) {
    std::cout << "self can't verify" << hashed_password;
    return false;
}
std::cout << "OK -> password and hash successfully verified";
return true;
```

The result is always false, here is an example from the console:  
```bash
self can't verify $argon2id$v=19$m=1048576,t=4,p=1$o8pnPhwRfQJcY/zCYpLQXg$/Q/D8qGFesHy6Y2xErDjE2rHfklUQVBE0blyDFals0o
```

crypto_secretbox_easy and crypto_secretbox_open_easy work good, am I missing any configuration?

**Comment by jedisct1** [comment]
The `SENSITIVE` settings may be a bit excessive for an iOS app. Can you try with more reasonable settings?

**Comment by jedisct1** [comment]
The `$argon2id$v=19$m=1048576,t=4,p=1$o8pnPhwRfQJcY/zCYpLQXg$/Q/D8qGFesHy6Y2xErDjE2rHfklUQVBE0blyDFals0o` indeed isn't valid for that example password.

I'll see if that can be reproduced via a `swift-sodium` test case.

Did you try with different Xcode versions? It may be an Xcode regression; that would not be the first one. That kind of bug is not fun at all to investigate, so if you need this urgently, I can do it under a contracting agreement. Otherwise, I won't have time to look at this until next year.

**Comment by jedisct1** [comment]
I was not able to replicate this with the simulator. Does it only happen on a device?

**Reply by jdelromano** [reply]
Hi, thank you for replying.

I was trying to adapt an existing app to work on iOS. I got troubles with the results of crypto_pwhash. But I got suspicious seeing that crypto_secretbox_easy and crypto_secretbox_open_easy were working fine.  
So to be sure I was not making something strange, I tried to run the example 2 on https://doc.libsodium.org/password_hashing/default_phf. 

It fails even using crypto_pwhash_OPSLIMIT_MODERATE and crypto_pwhash_MEMLIMIT_MODERATE.

I'm on Sequoia 15.1.1, XCode 16.1 and running on a device: iPad Pro 3rd gen (12.9 inch) ipadOS 18.0.1. I did not try on the simulator.

**Comment by jedisct1** [comment]
Can you try with the Clibsodium.xcframework files from https://github.com/jedisct1/swift-sodium ?

This is with `libsodium-stable`, right?

**Reply by jdelromano** [reply]
Yes, it's the stable released on the 4th december. 

No, it does not work.

**Reply by jdelromano** [reply]
I don't need it urgently, and probably for iOS and webassembly I will switch to libhydrogen.   
Since I had been working on it for a couple of days, I thought it was time to write.
